test/tools/test_tools_memories.py

```diff
@@ -58,3 +62,424 @@ def test_sopel_identifier_memory_channel_str():
+
+    memory.update(tuple_)
+    assert len(memory) == 2
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is True
+    assert 'froMlisTpairS' not in memory
+    assert 'froMdicT' not in memory
+    assert 'froMseT' not in memory
+
+    memory.update(list_)
+    assert len(memory) == 3
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is True
+    assert 'froMdicT' not in memory
+    assert 'froMseT' not in memory
+
+    memory.update(dict_)
+    assert len(memory) == 4
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is False
+    assert memory['froMdicT'] is True
+    assert 'froMseT' not in memory
+
+    memory.update(set_)
+    assert len(memory) == 5
+    assert memory['froMidentifieRmemorY'] is False
+    assert memory['froMtuplEpairS'] is False
+    assert memory['froMlisTpairS'] is False
+    assert memory['froMdicT'] is False
+    assert memory['froMseT'] is True
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|' operator added in py3.9")
+def test_sopel_identifier_memory_or_op():
+    memory = memories.SopelIdentifierMemory({
+        'FromMemory': True,
+        'FromOther': False,
+    })
+    assert len(memory) == 2
+    assert memory['froMmemorY'] is True
+    assert memory['froMotheR'] is False
+
+    other_memory = memories.SopelIdentifierMemory({
+        'FromMemory': False,
+        'FromOther': True,
+    })
+    assert len(other_memory) == 2
+    assert 'FromOther' in other_memory
+    assert other_memory['FromOther'] is True
+    assert 'fromother' in other_memory
+
+    result = memory | other_memory
+    assert len(result) == 2
+    assert result['fROmMemorY'] is False
+    assert result['fROmoTHEr'] is True
+
+    result = other_memory | memory
+    assert len(result) == 2
+    assert result['FroMMemorY'] is True
+    assert result['FroMOtheR'] is False
+
+
+# TODO: remove condition when dropping Python 3.8
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="'|' operator added in py3.9")
+def test_sopel_identifier_memory_or_op_dict():
+    memory = memories.SopelIdentifierMemory({'FromMemory': True, 'FromDict': False})
+    assert len(memory) == 2
+    assert memory['froMmemorY'] is True
+    assert memory['froMdicT'] is False
+
+    dictionary = {'FromMemory': False, 'FromDict': True, 'AlsoFromDict': True}
+    assert len(dictionary) == 3
+    assert dictionary['FromDict'] is True
+    assert dictionary['FromMemory'] is False
+    assert dictionary['AlsoFromDict'] is True
+    assert 'alsofromdict' not in dictionary
+
+    # __or__
+    result = memory | dictionary
+    assert len(result) == 3
+    assert result['fROmMemorY'] is False
+    assert result['fROmDicT'] is True
+    assert result['alsOfroMdicT'] is True
```

Comment:
Nit: this part of this test should still be run in Python 3.8, only `__ror__(dict)` is version-gated.

(probably the following clause can be wrapped in a version guard?)

Reply:
This appears to be a case where the nit is technically wrong (`__ror__` isn't version-gated) but it still pointed out something useful. It's true that you can't do `dict | dict` in Python 3.8, but you damn well can do:

- `SopelIdentifierMemory | SopelIdentifierMemory`
- `SopelIdentifierMemory | dict`
- `dict | SopelIdentifierMemory`

Plus the `|=` versions of those. Just having the relevant dunder methods makes it work. Next push will remove the version constraints from these tests entirely.